### PR TITLE
Cmake improvements and Coverage for C API

### PR DESCRIPTION
**Pull request by diptorupd**
Makes improvements to our Cmake scripts and moves all the CMake changes out of #241 to this PR.

Summary of changes:
- Moves all logic to search for dpcpp into a module called `FindDPCPP`.
- Adds a new option `DPCTL_GENERATE_COVERAGE` to enable coverage generation rules and dependencies. 
- Adds two rules (`llvm-cov` and `lcov-genhtml`) to generate code coverage for the C API when configured with `-DDPCTL_GENERATE_COVERAGE`.
    - Use `make llvm-cov` to generate the summary report for code coverage using `llvm-cov`. llvm-cov is part of standard LLVM tools and part of oneapi install.
    - Use `make lcov-genhtml` to generate a detailed html report for code coverage and missing segments of dpctl C API sources. Requires `lcov` to be installed on the system. Note that the new option `-DDPCTL_COVERAGE_REPORT_OUTPUT_DIR` can be used to configure where the html files are saved. If nothing is specified, the report is saved in the build directory in a subdirectory called `dpctl-c-api-coverage`.
- Other changes are level zero related that have been moved out of #241. 
    - A new module FindLevelZero checks if the level zero loader and headers are found on the build system.
    - A new option DPCTL_ENABLE_L0_PROGRAM_CREATION to toggle level zero program creation support. (Note the flag does nothing for now, as the related code changes are in #241)
- A generated configure file to store what options were used to build dpctl-CAPI. The file `dpctl_config.h` gets auto-generated in the install directory.

### CODE REVIEW

**Comment by diptorupd** [comment]
@PokhodenkoSA @reazulhoque @oleksandr-pavlyk Can you guys try out this PR? I will like to merge soon as my L0 work depends on the CMake changes here.

File: scripts/build_backend.py [hunk 1]
```diff
@@ -40,13 +40,13 @@
...
         "-DCMAKE_PREFIX_PATH=" + INSTALL_PREFIX,
-        "-DDPCPP_ROOT=" + DPCPP_ROOT,
+        "-DDPCPP_INSTALL_DIR=" + DPCPP_ROOT,
         "-DCMAKE_C_COMPILER:PATH=" + os.path.join(DPCPP_ROOT, "bin", "clang"),
         "-DCMAKE_CXX_COMPILER:PATH=" + os.path.join(DPCPP_ROOT, "bin", "clang++"),
         backends,
     ]
     subprocess.check_call(cmake_args, stderr=subprocess.STDOUT, shell=False)
-    subprocess.check_call(["make", "-j", "4"])
+    subprocess.check_call(["make", "V=1", "-j", "4"])
```
Comment (PokhodenkoSA) [inline scripts/build_backend.py:49]: V=1 - for verbose?

**Review by PokhodenkoSA — APPROVED** [review]
LGTM.

File: scripts/build_backend.py (hunk 1 above)
Comment (diptorupd) [reply scripts/build_backend.py:49]: Yes

**Pull request merged**